sycl-jit/jit-compiler/lib/rtc/DeviceCompilation.cpp

```diff
@@ -0,0 +1,111 @@
+//==---------------------- DeviceCompilation.cpp ---------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "DeviceCompilation.h"
+
+#include <clang/Basic/Version.h>
+#include <clang/CodeGen/CodeGenAction.h>
+#include <clang/Driver/Compilation.h>
+#include <clang/Frontend/CompilerInstance.h>
+#include <clang/Tooling/CompilationDatabase.h>
+#include <clang/Tooling/Tooling.h>
+
+#include <llvm/IR/Module.h>
+
+namespace {
+using namespace clang;
+using namespace clang::tooling;
+using namespace clang::driver;
+
+struct GetLLVMModuleAction : public ToolAction {
+  // Code adapted from `FrontendActionFactory::runInvocation`.
+  bool runInvocation(std::shared_ptr<CompilerInvocation> Invocation,
+                     FileManager *Files,
+                     std::shared_ptr<PCHContainerOperations> PCHContainerOps,
+                     DiagnosticConsumer *DiagConsumer) override {
+    assert(!Module && "Action should only be invoked on a single file");
+
+    // Create a compiler instance to handle the actual work.
+    CompilerInstance Compiler(std::move(PCHContainerOps));
+    Compiler.setInvocation(std::move(Invocation));
+    Compiler.setFileManager(Files);
+
+    // Create the compiler's actual diagnostics engine.
+    Compiler.createDiagnostics(DiagConsumer, /*ShouldOwnClient=*/false);
+    if (!Compiler.hasDiagnostics()) {
+      return false;
+    }
+
+    Compiler.createSourceManager(*Files);
+
+    // Ignore `Compiler.getFrontendOpts().ProgramAction` (would be `EmitBC`) and
+    // create/execute an `EmitLLVMOnlyAction` (= codegen to LLVM module without
+    // emitting anything) instead.
+    EmitLLVMOnlyAction ELOA;
+    const bool Success = Compiler.ExecuteAction(ELOA);
+    Files->clearStatCache();
+    if (!Success) {
+      return false;
+    }
+
+    // Take the module and its context to extend the objects' lifetime.
+    Module = ELOA.takeModule();
+    ELOA.takeLLVMContext();
+
+    return true;
+  }
+
+  std::unique_ptr<llvm::Module> Module;
+};
+
+} // anonymous namespace
+
+std::unique_ptr<llvm::Module> jit_compiler::compileDeviceCode(
+    const char *SYCLSource, View<IncludePair> IncludePairs,
+    View<const char *> UserArgs, const char *DPCPPRoot) {
+
+  SmallVector<std::string> CommandLine = {"-fsycl-device-only"};
+  // TODO: Allow instrumentation again when device library linking is
+  //       implemented.
+  CommandLine.push_back("-fno-sycl-instrument-device-code");
+  CommandLine.append(UserArgs.begin(), UserArgs.end());
+  clang::tooling::FixedCompilationDatabase DB{"./", CommandLine};
+
+  constexpr auto SourcePath = "rtc.cpp";
```

Comment:
rather than hard-code "rtc.cpp" in, can this 'fantasy name' for the file be an argument to the API?  It might show up in debug information, so it might be useful to users to be able to disambiguate, rather than having every dynamic device compiled kernel originate with the same fictional file name.

Reply:
Good idea, will do 👍

Reply:
Done. I'm passing in a semi-random ID, same as the file-based implementation. There's no property yet in the extension to specify a file name or prefix, correct?

Reply:
What would be the reason that a user might want to change this name?  Why is it better to have a semi-random ID as a default rather than a fixed string?

I can imagine that the name might show up in error / log messages, for example, when there is a syntax error in the source string.  If that's the only case the name is visible, it seems like having a fixed string like `rtc.cpp` would be fine, and probably preferable to a name with a random number.

I'm not opposed to adding a property which allows the user to set this name, but I think it should be an optional property because I think many people will not care what the name is.

Reply:
> I can imagine that the name might show up in error / log messages, for example, when there is a syntax error in the source string. If that's the only case the name is visible, [...]

Yes, that's the only case. I agree that the ID doesn't add much value here because we don't materialise anything on the actual filesystem. I'll keep the plumbing to pass the filename down to the JIT, but will set it `rtc.cpp` until there's a need and a means to modify it from the extension.